File: Fortran/IBM/execute/F2003/OO_procptr/bindc2/procptrBindcFunptr05.f90
!#######################################################################
!*  ===================================================================
!*  XL Fortran Test Case                          IBM INTERNAL USE ONLY
!*  ===================================================================
!*  ===================================================================
!*
!*  TEST CASE TITLE            :
!*
!*  PROGRAMMER                 : William Zhang 
!*  DATE                       : 3/01/2006
!*  ORIGIN                     : AIX Compiler Development, Toronto Lab
!*                             :
!*
!*  PRIMARY FUNCTIONS TESTED   : Procedure Pointer with BindC 
!*                             :
!*  SECONDARY FUNCTIONS TESTED : 
!*
!*
!*  DESCRIPTION                :  
!*                              function pointer which points to C function with int
!*                              argument and void return, as actual argument 
!*                              passed directly to C  through procedure
!*                              pointer reference. 
!* ===================================================================

module fptr05
   use ISO_C_BINDING
   interface
       subroutine csub(i, cptr) bind(c)
          import C_INT, C_FUNPTR
          integer(C_INT) :: i
          type(C_FUNPTR) :: cptr
       end subroutine csub
   end interface
end module fptr05

program procptrBindcFunptr05

   use ISO_C_BINDING
  
   use fptr05

   interface
       subroutine swap(i, j) bind(c)
          import C_INT
          integer(C_INT) :: i, j
       end subroutine 
   end interface

   integer(C_INT) :: i 

   procedure(csub), pointer :: subind 

   subind => csub   

   i = 34_C_INT

   call subind(i, C_FUNLOC(swap))

   if(i .ne. 22) error stop 1_4 

end program procptrBindcFunptr05

File: Fortran/IBM/execute/F2003/OO_procptr/bindc2/procptrBindcFunptr05.c

#include <stdio.h>
#include <stdlib.h>

void csub(int *p, void (**f)(int * , int *)) {

   int a;
   int b;

   a = 14;
   b = 8;

   if ( *p != 34 ) exit(41);
   if( a != 14 ) exit(42);
   if( b != 8 ) exit(43);

   (*f)(&a, &b); 

   *p = 22;

   if( a != 8 ) exit(44);
   if( b != 14 ) exit(45);

}

void  swap(int *p1, int *p2) {

   int tmp;

   if (*p1 != 14 ) exit(46);  
   if (*p2 != 8 ) exit(47);

   tmp = *p2;
   *p2 = *p1;
   *p1 = tmp;

}

